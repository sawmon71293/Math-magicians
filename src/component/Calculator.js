import { useReducer } from 'react';
import { DigitButton, ACTIONS } from './DigitButton';
import OperationButton from './OperationButton';
import evaluate from './Evaluate';

// eslint-disable-next-line consistent-return
function reducer(state, { type, payload }) {
  // eslint-disable-next-line default-case
  switch (type) {
    case ACTIONS.ADD_DIGIT:
      if (state.overwrite) {
        return {
          ...state,
          currentOperand: payload.digit,
          overwrite: false,
        };
      }
      if (payload.digit === '0' && state.currentOperand === '0') return state;
      if (payload.digit === '.' && state.currentOperand.inclued('.')) {
        return state;
      }
      return {
        ...state,
        currentOperand: `${state.currentOperand || ''}${payload.digit}`,
      };
    case ACTIONS.CHOOSE_OPERATION:
      if (state.currentOperand == null && state.previousOperand == null) {
        return state;
      }
      if (state.currentOperand == null) {
        return {
          ...state,
          operation: payload.operation,
        };
      }

      if (state.previousOperand == null) {
        return {
          ...state,
          operation: payload.operation,
          previousOperand: state.currentOperand,
          currentOperand: null,
        };
      }

      return {
        ...state,
        previousOperand: evaluate(state),
        operation: payload.operation,
        currentOperand: null,
      };
    case ACTIONS.CLEAR:
      return {};
    case ACTIONS.EVALUATE:
      if (state.operation == null
          || state.currentOperand == null
          || state.previousOperand == null) {
        return state;
      }
      return {
        ...state,
        overwrite: true,
        previousOperand: null,
        operation: null,
        currentOperand: evaluate(state),
      };
  }
}

export default function Calculator() {
  const [{ currentOperand, previousOperand, operation }, dispatch] = useReducer(
    reducer,
    {},
  );
  return (
    <div className="wrapper">
      <div className="screen">
        <div>
          {previousOperand}
          {' '}
          {operation}
          {' '}
          {currentOperand}
        </div>

      </div>
      <button
        type="button"
        className="btn light-gray"
        onClick={() => dispatch({ type: ACTIONS.CLEAR })}
      >
        AC
      </button>
      <button
        className="btn light-gray"
        type="button"
        onClick={() => dispatch({ type: ACTIONS.CHOOSE_OPERATION })}
      >
        +/-
      </button>
      <button
        className="btn light-gray"
        type="button"
        onClick={() => dispatch({ type: ACTIONS.CHOOSE_OPERATION })}
      >
        %
      </button>
      <OperationButton
        operation="รท"
        dispatch={dispatch}
      />
      <DigitButton
        digit="7"
        dispatch={dispatch}
      />
      <DigitButton
        digit="8"
        dispatch={dispatch}
      />
      <DigitButton
        digit="9"
        dispatch={dispatch}
      />
      <OperationButton
        operation="x"
        dispatch={dispatch}
      />
      <DigitButton
        digit="4"
        dispatch={dispatch}
      />
      <DigitButton
        digit="5"
        dispatch={dispatch}
      />
      <DigitButton
        digit="6"
        dispatch={dispatch}
      />
      <OperationButton
        operation="-"
        dispatch={dispatch}
      />
      <DigitButton
        digit="1"
        dispatch={dispatch}
      />
      <DigitButton
        digit="2"
        dispatch={dispatch}
      />
      <DigitButton
        digit="3"
        dispatch={dispatch}
      />
      <OperationButton
        operation="+"
        dispatch={dispatch}
      />
      <DigitButton
        digit="0"
        dispatch={dispatch}
      />
      <DigitButton
        digit="."
        className="span-two"
        dispatch={dispatch}
      />
      <button
        className="btn orange"
        type="button"
        onClick={() => dispatch({ type: ACTIONS.EVALUATE })}
      >
        =
      </button>
    </div>
  );
}
