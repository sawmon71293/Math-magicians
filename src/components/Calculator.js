import { useState } from 'react';
import calculate from '../logic/Calculate';
import BUTTONS from './Buttons';

export default function Calculator() {
  const [data, setData] = useState({ total: null, next: null, operator: null });
  const handleClick = (buttonName) => {
    const newData = calculate(data, buttonName);
    setData(newData);
  };

  const renderButtons = (button) => (
    <button
      key={button.name}
      className={`btn ${button.type}`}
      type="button"
      onClick={() => { handleClick(button.name); }}
    >
      {button.name}
    </button>
  );

  return (
    <section className="container calculator">
      <div>
        <h3>Let&apos;s do some Math</h3>
      </div>
      <div className="wrapper">
        <div className="screen">
          <div>
            {data.next || data.total || '0'}
            {data.operator && ` ${data.operator}`}
          </div>
        </div>
        {BUTTONS.map(renderButtons)}
      </div>
    </section>
  );
}
