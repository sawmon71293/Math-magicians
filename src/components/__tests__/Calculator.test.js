import React from 'react';
import { render } from '@testing-library/react';
import Calculator from '../Calculator';
import '@testing-library/jest-dom/extend-expect';

describe('Calculator component', () => {
  it('renders initial state correctly', () => {
    const { getAllByText } = render(<Calculator />);

    // Assert that the initial state is rendered correctly
    expect(getAllByText("Let's do some Math")).toBeInTheDocument();
    expect(getAllByText('0').length).toBe(2);
  });
});
