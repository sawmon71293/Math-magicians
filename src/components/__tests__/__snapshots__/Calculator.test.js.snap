// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component renders correctly 1`] = `
<section
  className="container calculator"
>
  <div>
    <h3>
      Let's do some Math
    </h3>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="screen"
    >
      <div>
        0
      </div>
    </div>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      className="btn light-gray span-two"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</section>
`;

exports[`Calculator component renders correctly with specific props 1`] = `
<section
  className="container calculator"
>
  <div>
    <h3>
      Let's do some Math
    </h3>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="screen"
    >
      <div>
        0
      </div>
    </div>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      className="btn light-gray span-two"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      className="btn light-gray"
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="btn orange"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</section>
`;
